h2olog: drop the root privilege once BPF is set up

toru opened the pull request:
Technically `h2olog` only needs root privilege to:

1. Attach the BPF tracepoints
2. Open the BPF ring buffer

Therefore drop the root privilege as soon as the two operations are complete. Error handling may seem unforgiving but it follows a consistent pattern in `h2olog`.

gfx commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +125,34 @@ static void show_process(pid_t pid)
+            exit(EXIT_FAILURE);
+        }
+        uid_t uid = atoi(sudo_uid);
+        if (uid > 0 && setuid(uid) != 0) {
```
```suggestion
        if (uid != 0 && setuid(uid) != 0) {
```

toru replied:
Thank you @gfx. This is addressed in: df7ba942

kazuho reviewed (commented):
Thank you for the pull request. I think this is a nice change; left some thoughts. PTAL.

gfx commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +125,44 @@ static void show_process(pid_t pid)
 
+static void drop_root_privilege(void)
+{
+    if (getgid() == 0) {
```
In addition, h2olog already checks euid at the startup: https://github.com/h2o/h2o/blob/34e41a2013dbdfaed776fcc71e64962b1b207d9a/src/h2olog/h2olog.cc#L239-L242

toru replied:
Thanks you two. The `setgid()` code is now nested under `if (getuid() == 0)` - e85fc361

toru commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +125,44 @@ static void show_process(pid_t pid)
+        const char *sudo_gid = getenv("SUDO_GID");
+        if (sudo_gid == NULL) {
+            fprintf(stderr, "Error: failed to read the SUDO_GID env variable\n");
+            exit(EXIT_FAILURE);
```
Thank you for pointing out my blind spot. Yes, adding a command line option to skip privdrop makes sense to me. This is addressed in: ce5e7b6 and 35ce6ca

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -190,12 +229,13 @@ int main(int argc, char **argv)
     }
 
     int debug = 0;
+    int drop_root = 1;
```
I think it might be less confusing if we rename this to `preserve_root` or something to avoid using inverted definitions against the command line option.

toru replied:
Nice! yes, thank you for pointing this out. This definitely requires few brain cycles to grok.

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,44 @@ static void show_process(pid_t pid)
+static void drop_root_privilege(void)
+{
+    if (getuid() == 0) {
+        if (getgid() == 0) {
```
I think it might be possible remove this `if`, is it necessary? I think that the behavior we want is "if running as root, drop root privileges."

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,44 @@ static void show_process(pid_t pid)
+            errno = 0;
+            gid_t gid = (gid_t)strtol(sudo_gid, NULL, 10);
+            if (errno != 0) {
+                fprintf(stderr, "Error: overflow while parsing SUDO_GID\n");
```
```suggestion
                fprintf(stderr, "Error: failed to parse SUDO_GID\n");
```
... or something alike. I think that the parsing might fail for reasons other than overflow?

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,44 @@ static void show_process(pid_t pid)
+                fprintf(stderr, "Error: overflow while parsing SUDO_GID\n");
+                exit(EXIT_FAILURE);
+            }
+            if (gid != 0 && setgid(gid) != 0) {
```
I think we'd either remove `gid != 0` or call it an error.

Though it depends on how we define what "dropping root privileges" mean. If we are to define the act as switching to SUDO_UID / SUDO_GID, then simply removing this condition is fine. If we are to define the act as "dropping root privileges" in a strict manner, then we should call it an error.

toru replied:
Agreed that we need to decide on the definition, and I'm good with simply: "switching to SUDO_UID / SUDO_GID" as you mentioned. I see your point about needing to be more unforgiving if we had chosen a stricter definition. Thanks!

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,44 @@ static void show_process(pid_t pid)
+        errno = 0;
+        uid_t uid = (uid_t)strtol(sudo_uid, NULL, 10);
+        if (errno != 0) {
+            fprintf(stderr, "Error: overflow while parsing SUDO_UID\n");
```
Please see my comment above.

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,44 @@ static void show_process(pid_t pid)
+            fprintf(stderr, "Error: overflow while parsing SUDO_UID\n");
+            exit(EXIT_FAILURE);
+        }
+        if (uid != 0 && setuid(uid) != 0) {
```
Here also.

kazuho reviewed (commented):
Thank you for the changes. Left some comments.

toru commented:
@kazuho As always, thank you for the thorough review. I've addressed your feedback in the following commits:

108bdbb - rename the drop_root option var to preserve_root
2e9fa21 - update the error message
7937ab9 -  call setuid() and setgid() regardless of the gid/uid

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,42 @@ static void show_process(pid_t pid)
+            exit(EXIT_FAILURE);
+        }
+        if (setgid(gid) != 0) {
+            fprintf(stderr, "Error: failed to drop the root group\n");
```
I think we might want to use `perror` for syscalls, because the cause of the failure could be anything.

toru replied:
Thank you for noticing. This did not occur to me at the time. Addressed in: e2aef5e

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,42 @@ static void show_process(pid_t pid)
+            exit(EXIT_FAILURE);
+        }
+        if (setuid(uid) != 0) {
+            fprintf(stderr, "Error: failed to drop the root user\n");
```
Same here.

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,42 @@ static void show_process(pid_t pid)
+    if (getuid() == 0) {
+        const char *sudo_gid = getenv("SUDO_GID");
+        if (sudo_gid == NULL) {
+            fprintf(stderr, "Error: failed to read the SUDO_GID env variable\n");
```
Should we better say that the environment variable is not set, rather than "failed to read"?

toru replied:
Great catch! Addressed in: b6c5d3f

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,42 @@ static void show_process(pid_t pid)
+        }
+        const char *sudo_uid = getenv("SUDO_UID");
+        if (sudo_uid == NULL) {
+            fprintf(stderr, "Error: failed to read the SUDO_UID env variable\n");
```
Same here.

kazuho reviewed (commented):
Thank you for the changes. LGTM (with minor suggestions).

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,42 @@ static void show_process(pid_t pid)
+            exit(EXIT_FAILURE);
+        }
+        if (setgid(gid) != 0) {
+            perror("Error: setgid(2) failed\n");
```
```suggestion
            perror("Error: setgid(2) failed");
```

perror will append a colon, the error message, and a newline.

kazuho commented on src/h2olog/h2olog.cc:
```diff
@@ -125,6 +126,42 @@ static void show_process(pid_t pid)
+            exit(EXIT_FAILURE);
+        }
+        if (setuid(uid) != 0) {
+            perror("Error: setuid(2) failed\n");
```
```suggestion
            perror("Error: setuid(2) failed");
```
